keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,551 @@
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+In the JobSet API, ReplicatedJobs are the individual Job elements that are created.  These are Replicated so you can have the same Job replicated.  This adds some complexity in the concept of Ready as we must consider the overall ReplicatedJob Ready rather than using the individual Jobs.
+
+For example, a Job with a `Parallelism`: 4 and `Ready` of 3. Means that 1 pod of a Job is not ready.  
+The Job API has no concept of an overall readiness condition based on a percentage of pods that are ready.  
+
+We could 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+
+```golang
+// Status of ReplicatedJob for StartupPolicy
+type Status string
+
+const (
+ // Ready means that the job is ready
+ Ready Status = "Ready"
+
+ // Succeeded means that the job is successful
+ Succeeded Status = "Succeeded"
+)
+
+type StartupPolicy struct {
+ // TargetReplicatedJobStatuses is a list of rules for applying StartupPolicy.
+ // +optional
+ // +listType=map
+ // +listMapKey=name
+ TargetReplicatedJobStatuses []TargetReplicatedJobStartup `json:"targetReplicatedJobStatuses,omitempty"`
+}
+
+type TargetReplicatedJobStartup struct {
+ // Name of ReplicatedJob to apply startup policy
+ Name string `json:"name"`
+
+ // The status you want the ReplicatedJob to be
+ // +kubebuilder:validation:Enum=Ready;Succeeded
+ Status Status `json:"status"`
+
+ // Number of Statuses is how many ReplicatedJob statuses to consider "started"
+ // Empty will mean all status must match the replicas.
+ NumberOfStatuses int32 `json:"numberOfStatuses,omitempty"`
+}
+```
+
+Keeping the non specified jobs halted we propose to use suspend for these jobs.
+
+So we will add a new field to the ReplicatedJobStatus.
+
+```golang
+// ReplicatedJobStatus defines the observed ReplicatedJobs Readiness.
+type ReplicatedJobStatus struct {
+  ...
+	Suspended int32  `json:"suspended"`
+}
+```
+
+### Implementation
+
+If StartupPolicy is set and suspend field is off, we will start all jobs in a suspended state.  We will not set a condition for JobSet as suspended.
+
+To track which ReplicatedJobs are suspended, we will add a `Suspended` field to the ReplicatedJobStatus.  This field will track the number of replicas that are suspended.
+
+When a JobSet is started with `StartupPolicy` specified, we will create jobs in a suspended state (ie `Job.Spec.Suspend = true`)  This avoids starting the underlying jobs.
```

Comment:
Correct me if I am wrong but I think this saying the same thing as the first sentence in the implementation section?

Reply:
you are correct. updated